sched/semaphore/sem_holder.c

```diff
@@ -880,7 +880,7 @@ void nxsem_canceled(FAR struct tcb_s *stcb, FAR sem_t *sem)
 {
   /* Check our assumptions */
 
-  DEBUGASSERT(atomic_load(NXSEM_COUNT(sem)) <= 0);
+  DEBUGASSERT((int16_t)atomic_load(NXSEM_COUNT(sem)) <= 0);
```

Comment:
why need cast if NXSEM_COUNT return atomic_short?

Reply:
atomic_load returns uint64_t.

Reply:
see
https://github.com/apache/nuttx/blob/daab676db97bbf2690e9011cbed1a0d45b2112f6/include/nuttx/lib/stdatomic.h#L75-L81
https://github.com/apache/nuttx/blob/daab676db97bbf2690e9011cbed1a0d45b2112f6/include/nuttx/lib/stdatomic.h#L201-L204

Reply:
but the spec requires it return the same base type:
https://en.cppreference.com/w/c/atomic/atomic_load
we should fix the implementation instead.

Reply:
IMO, we should not pretend to have generic selection.
it's simpler to use concrete-type apis like, say, nx_atomic_load_int16.

Reply:
anyway, while it's fine to have a discussion about how it should be, the breakage is too severe to leave during the discussion, IMO.
please fix it soon, either with this PR or by reverting the changes in question.

Reply:
> IMO, we should not pretend to have generic selection. it's simpler to use concrete-type apis like, say, nx_atomic_load_int16.

Yes, but it's this is the standard defined prototype:(

Reply:
@xiaoxiang781216  But we cannot make a c function's return value support both signed and unsigned numbers simultaneously.

Reply:
Yes, we need macro here.

Reply:
> @crafcat7 please fix this ASAP.

Later we can consider whether there is a better way to reorganize arch_atomic & nuttx/stdatomic.c so that they can directly go to the atomic processing of the corresponding type, in other words, it can return the same result type as the input type parameter.
This work should take some time

Reply:
> > IMO, we should not pretend to have generic selection. it's simpler to use concrete-type apis like, say, nx_atomic_load_int16.
> 
> Yes, but it's this is the standard defined prototype:(

my suggestion is to use nuttx-specific, non-standard prototype.

besides that, we can provide c11 stdatomic to user applications where possible.
but our own code (eg. semaphore implementation) should not use it, IMO.

Reply:
>  This work should take some time.

that's my expection too.

in the meantime, IMO, we should make band-aid fixes (like this PR) or a revert.

Reply:
Ok, let's limit the kernel to use only one type of atomic type, so we can provide an implementation when the compiler doesn't provide the atomic operation.

Reply:
i'd suggest to revert the change in question for now because it would take some time to fix regressions: https://github.com/apache/nuttx/pull/14804